### PR TITLE
Operation Collections in Lab

### PR DESCRIPTION
**Pull request by B2o5T**
origin PR https://github.com/kamilkisiela/graphql-hive/pull/1558

### CODE REVIEW

File: packages/services/api/src/modules/collection/module.graphql.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    headers: JSON
+  }
+
+  extend type Mutation {
+    createOperation(input: CreateOperationInput!): Operation!
+    updateOperation(input: UpdateOperationInput!): Operation!
+    deleteOperation(id: ID!): Operation!
+    createCollection(input: CreateCollectionInput!): Collection!
+    updateCollection(input: UpdateCollectionInput!): Collection!
+    deleteCollection(id: ID!): Collection!
```
Comment (n1ru4l) [inline packages/services/api/src/modules/collection/module.graphql.ts:74]: I feel like these field names should be more detailed, e.g. `createOperationForCollection` or `createDocumentCollectionOperation`, same for the object types. 🤔

File: packages/services/api/src/modules/collection/module.graphql.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  extend type Query {
+    collections(targetId: ID!): CollectionConnection!
+    operation(id: ID!): Operation!
+    collection(id: ID!): Collection!
+  }
```
Comment (n1ru4l) [inline packages/services/api/src/modules/collection/module.graphql.ts:81]: Why expose these on the `Query` root type when they are tied to a target?

`Target.collections(first:,after:)`, `Collection.operation(id:)`, `Target.collection(id:)`, makes more sense to me.

File: packages/services/api/src/modules/collection/providers/tokens.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { InjectionToken } from 'graphql-modules';
+
+export type CollectionConfig = {
+  // endpoint: string | null;
```
Comment (n1ru4l) [inline packages/services/api/src/modules/collection/providers/tokens.ts:4]: Why is this commented out, do we need this?

File: packages/web/app/src/components/v2/modals/delete-collection.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import { ReactElement } from 'react';
+import { useMutation } from 'urql';
+import { Button, Heading, Modal } from '@/components/v2';
+import { DeleteCollectionDocument } from '@/graphql';
```
Comment (n1ru4l) [inline packages/web/app/src/components/v2/modals/delete-collection.tsx:4]: Why avoid using client preset?

**Review by n1ru4l — COMMENTED** [review]
added some comments

File: packages/web/app/src/components/v2/modals/create-collection.tsx [hunk 5]
```diff
@@ -0,0 +1,161 @@
+import { ReactElement, useEffect } from 'react';
+import { useFormik } from 'formik';
+import { useMutation, useQuery } from 'urql';
+import * as Yup from 'yup';
+import { Button, Heading, Input, Modal } from '@/components/v2';
+import {
+  CollectionDocument,
+  CreateCollectionDocument,
+  TargetDocument,
+  UpdateCollectionDocument,
```
[the remark this answers is not in the document]
Comment (dotansimha) [reply packages/web/app/src/components/v2/modals/create-collection.tsx:10]: Agree, @B2o5T let's switch to use the client preset

File: packages/services/api/src/modules/collection/module.graphql.ts (hunk 2 above)
Comment (dotansimha) [reply packages/services/api/src/modules/collection/module.graphql.ts:81]: Yeah I agree, let's change that @B2o5T

File: packages/services/api/src/modules/collection/resolvers.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-nocheck TODO: fix Property 'node' does not exist on type ...
```
Comment (B2o5T) [inline packages/services/api/src/modules/collection/resolvers.ts:2]: can someone fixes type errors here?

File: packages/services/storage/src/index.ts [hunk 7]
```diff
@@ -3145,10 +3141,6 @@ export async function createStorage(connection: string, maximumPoolSize: number)
           , to_json("updated_at") as "updatedAt"
       `);
 
-      if (result == null) {
-        return null;
-      }
```
Comment (n1ru4l) [inline packages/services/storage/src/index.ts]: Why do you prefer to panic and raise an exception?

File: packages/services/api/src/modules/collection/providers/collection.provider.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  async createOperation(input: CreateDocumentCollectionOperationInput) {
+    const currentUser = await this.authManager.getCurrentUser();
+    return this.storage.createDocumentCollectionDocument({
+      documentCollectionId: input.collectionId,
+      title: input.name,
+      contents: input.query,
+      variables: input.variables,
+      headers: input.headers,
```
Comment (n1ru4l) [inline packages/services/api/src/modules/collection/providers/collection.provider.ts:69]: I think we should validate the input and set maximum lengths before inserting the data into the database? Errors can be exposed using the output object type ok error pattern

Comment (n1ru4l) [reply packages/services/api/src/modules/collection/providers/collection.provider.ts:69]: (same applies for all other mutation fields added in this PR)

File: packages/services/api/src/modules/shared/providers/storage.ts [hunk 9]
```diff
@@ -659,15 +659,13 @@ export interface Storage {
     variables: string | null;
     headers: string | null;
     createdByUserId: string | null;
-  }): Promise<DocumentCollectionDocument>;
+  }): Promise<DocumentCollectionOperation>;
 
   /**
    * Returns null if the document collection document does not exist (did not get deleted).
    * Returns the id of the deleted document collection document if it got deleted
    */
```
[the remark this answers is not in the document]
Comment (dotansimha) [reply packages/services/api/src/modules/shared/providers/storage.ts]: Yeah agree. I don't like the SQL errors but it seems fine for the time being. I mean, returning `null` and then doing a check to return "collection not found" is not that different than having "resource not found".

File: packages/services/api/src/modules/collection/module.graphql.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+  type DocumentCollectionConnection {
+    nodes: [DocumentCollection!]!
+    total: Int!
+  }
+
+  type DocumentCollectionOperationsConnection {
+    nodes: [DocumentCollectionOperation!]!
+    total: Int!
+  }
```
Comment (n1ru4l) [inline packages/services/api/src/modules/collection/module.graphql.ts:22]: https://relay.dev/graphql/connections.htm 👀

Comment (B2o5T) [reply packages/services/api/src/modules/collection/module.graphql.ts:22]: please be more descriptive, do you mean `total` field?

Comment (n1ru4l) [reply packages/services/api/src/modules/collection/module.graphql.ts:22]: Please follow the connection specification for naming and implementing Connection fields and types as we already do for all new fields added.

`nodes` should be `edges` and the `PageInfo` field should be added.

Comment (B2o5T) [reply packages/services/api/src/modules/collection/module.graphql.ts:22]: I am not sure if it was me or @dotansimha who made yesterday refactoring previously we had `OperationEdge`. This is a nice time to extend `graphql-eslint` relay config hehe

File: packages/services/api/src/modules/shared/providers/storage.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (dotansimha) [reply packages/services/api/src/modules/shared/providers/storage.ts]: Ok will fix

File: integration-tests/tests/storage/document-collection.spec.ts [hunk 11]
```diff
@@ -0,0 +1,487 @@
+/* eslint-disable no-process-env */
```
[the remark this answers is not in the document]
Comment (dotansimha) [reply integration-tests/tests/storage/document-collection.spec.ts:1]: fixing

**Review by dotansimha — APPROVED** [review]

**Review by kamilkisiela — APPROVED** [review]

**Pull request merged**